src/retry_logic.rs

```diff
@@ -0,0 +1,110 @@
+use std::time::Duration;
+use temporal_sdk_core_protos::coresdk::common::RetryPolicy;
+
+pub(crate) trait RetryPolicyExt {
+    /// Ask this retry policy if a retry should be performed. Caller provides the current attempt
+    /// number - the first attempt should start at 1.
+    ///
+    /// Returns `None` if it should not, otherwise a duration indicating how long to wait before
+    /// performing the retry.
+    fn should_retry(&self, attempt_number: usize, err_str: &str) -> Option<Duration>;
+}
+
+impl RetryPolicyExt for RetryPolicy {
+    fn should_retry(&self, attempt_number: usize, err_str: &str) -> Option<Duration> {
+        if attempt_number >= self.maximum_attempts.max(0) as usize {
+            return None;
+        }
+
+        for pat in &self.non_retryable_error_types {
+            if err_str.contains(pat.as_str()) {
+                return None;
+            }
+        }
+
+        let converted_interval = self
+            .initial_interval
+            .clone()
+            .map(TryInto::try_into)
+            .transpose()
+            .ok()
+            .flatten();
+        if attempt_number == 1 {
+            return converted_interval;
+        }
+        let coeff = if self.backoff_coefficient != 0. {
+            self.backoff_coefficient
+        } else {
+            2.0
+        };
```

Comment:
I kinda understand wanting the other single statement to become a single-statement method since it has reuse potential (even then single statements with hardly any logic rarely deserve to be broken out, otherwise one could only imagine the extent which that could be applied). But for this simple if statement I, subjectively, prefer it remain inline.

Reply:
Yeah, not gonna change this one. This would be like making a fn for a ternary op